### PR TITLE
allow different base_elements for compositional fields

### PR DESCRIPTION
**Pull request by tjhei**
We now store different base_element indices in introspection (one index for each compositional field). This is no functional change, as we do not support different FiniteElements so far.

To be able to make this change, I had to fix various places that use base_elements or assume that all compositional fields have the same base element.

part of #5748

### CODE REVIEW

**Comment by tjhei** [comment]
This one if for you @bangerth . :-)

**Comment by bangerth** [comment]
There are numerous errors such as these:
```
In file included from /Users/jenkins/jenkins/workspace/aspect-osx_PR-5802/build/CMakeFiles/aspect-release.dir/Unity/unity_40_cxx.cxx:16:
[1m/Users/jenkins/jenkins/workspace/aspect-osx_PR-5802/source/simulator/helper_functions.cc:1690:74: [0m[0;1;31merror: [0m[1mno viable conversion from 'const std::vector<unsigned int>' to 'unsigned int'[0m
    const Quadrature<dim> quadrature_C(dof_handler.get_fe().base_element(introspection.base_elements.compositional_fields).get_unit_support_points());
[0;1;32m                                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[0m[1m/Users/jenkins/jenkins/workspace/aspect-osx_PR-5802/source/simulator/helper_functions.cc:2578:22: [0m[0;1;30mnote: [0min instantiation of member function 'aspect::Simulator<2>::compute_reactions' requested here[0m
  ASPECT_INSTANTIATE(INSTANTIATE)
[0;1;32m                     ^
[0m[1m/Users/jenkins/candi-9.5.1-r1b/deal.II-v9.5.1/include/deal.II/fe/fe.h:1622:35: [0m[0;1;30mnote: [0mpassing argument to parameter 'index' here[0m
  base_element(const unsigned int index) const;
[0;1;32m                                  ^
[0mIn file included from /Users/jenkins/jenkins/workspace/aspect-osx_PR-5802/build/CMakeFiles/aspect-release.dir/Unity/unity_40_cxx.cxx:16:
[1m/Users/jenkins/jenkins/workspace/aspect-osx_PR-5802/source/simulator/helper_functions.cc:1785:74: [0m[0;1;31merror: [0m[1mno viable conversion from 'const std::vector<unsigned int>' to 'unsigned int'[0m
              for (unsigned int j=0; j<dof_handler.get_fe().base_element(introspection.base_elements.compositional_fields).dofs_per_cell; ++j)
[0;1;32m                                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[0m[1m/Users/jenkins/candi-9.5.1-r1b/deal.II-v9.5.1/include/deal.II/fe/fe.h:1622:35: [0m[0;1;30mnote: [0mpassing argument to parameter 'index' here[0m
  base_element(const unsigned int index) const;
[0;1;32m                                  ^
[0mIn file included from /Users/jenkins/jenkins/workspace/aspect-osx_PR-5802/build/CMakeFiles/aspect-release.dir/Unity/unity_40_cxx.cxx:16:
[1m/Users/jenkins/jenkins/workspace/aspect-osx_PR-5802/source/simulator/helper_functions.cc:1826:70: [0m[0;1;31merror: [0m[1mno viable conversion from 'const std::vector<unsigned int>' to 'unsigned int'[0m
          for (unsigned int j=0; j<dof_handler.get_fe().base_element(introspection.base_elements.compositional_fields).dofs_per_cell; ++j)
[0;1;32m                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
I would assume that you get those on your machine too?

File: include/aspect/introspection.h [hunk 1]
```diff
@@ -465,6 +465,21 @@ namespace aspect
        * @}
        */
 
+      /**
+       * Return a vector with all used indices of base elements of the deal.II FiniteElement
+       * that are used for compositional fields. If several fields are the same type, they
+       * share base elements.
+      */
+      std::vector<unsigned int>
+      get_compositional_field_base_element_indices() const;
```
Comment (bangerth) [inline include/aspect/introspection.h:474]: Am I correct that the output array has length `# of compositional fields`? If so, would you mind adding that.

File: include/aspect/introspection.h [hunk 2]
```diff
@@ -465,6 +465,21 @@ namespace aspect
...
+
+      /**
+       * Return a vector with all compositional field indices that belong to a given
+       * base element index.
+      */
+      std::vector<unsigned int>
+      get_compositional_field_indices_with_base_element(const unsigned int base_element_index) const;
```
Comment (bangerth) [inline include/aspect/introspection.h:481]: Is this indexed starting at the zeroth compositional element? Or at the velocity? 

Or perhaps you are saying that the argument must be an element of the array that `get_compositional_field_base_element_indices()` returns? If so, please add that. (And assert?)

File: source/mesh_refinement/compaction_length.cc [hunk 3]
```diff
@@ -65,7 +70,7 @@ namespace aspect
                         ExcMessage("Need MeltOutputs from the material model for computing the melt properties."));
 
             // for each composition dof, check if the compaction length exceeds the cell size
-            for (unsigned int i=0; i<this->get_fe().base_element(this->introspection().base_elements.compositional_fields).dofs_per_cell; ++i)
+            for (unsigned int i=0; i<in.n_evaluation_points(); ++i)
```
Comment (bangerth) [inline source/mesh_refinement/compaction_length.cc:73]: I would accept this commit separately if it's cumbersome to carry around.

File: source/mesh_refinement/composition_gradient.cc [hunk 4]
```diff
@@ -40,41 +41,48 @@ namespace aspect
...
-                               this->get_fe(),
-                               quadrature,
-                               update_quadrature_points | update_gradients);
-
-      // the values of the compositional fields are stored as block vectors for each field
-      // we have to extract them in this structure
-      std::vector<Tensor<1,dim>> composition_gradients (quadrature.size());
-
-      for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
+      for (const unsigned int k : this->introspection().get_compositional_field_base_element_indices())
```
Comment (bangerth) [inline source/mesh_refinement/composition_gradient.cc:44]: This is a long loop body to carry around a single-character variable name that is not standard (like `c` for composition). Perhaps name it `base_element_index` or some such?

**Review by bangerth — COMMENTED** [review]
First part.

File: source/simulator/introspection.cc [hunk 5]
```diff
@@ -487,6 +520,22 @@ namespace aspect
...
 
+
+  template <int dim>
+  bool
+  Introspection<dim>::is_composition_component (const unsigned int component_index) const
+  {
+    // All compositions live at the end. Just to be sure, there are no other components
+    // in our system after compositional fields, right?
+    Assert(component_indices.compositional_fields[0] > component_indices.temperature
+           && (*component_indices.compositional_fields.end()) == n_components-1, ExcInternalError());
```
Comment (bangerth) [inline source/simulator/introspection.cc:531]: Wait what? You're dereferencing the `end()` iterator? I must be missing something, or this is not likely going to go well...

File: source/simulator/introspection.cc [hunk 6]
```diff
@@ -332,6 +332,39 @@ namespace aspect
...
+  }
+
+
+
+  template <int dim>
+  std::vector<unsigned int>
+  Introspection<dim>::get_compositional_field_indices_with_base_element(const unsigned int base_element_index) const
+  {
+    std::vector<unsigned int> result;
+
```
Comment (bangerth) [inline source/simulator/introspection.cc:357]: This is where you want to check that `base_element_index` is ever used. You can probably do that by making sure that the output vector is non-empty.

File: source/simulator/introspection.cc [hunk 7]
```diff
@@ -332,6 +332,39 @@ namespace aspect
...
+    // We are assigning base elements in order, so the first compositional field
+    // gives us the first base element index. Then we find the largest index
+    // in the vector. This is necessary, because the fields could have type A,B,A.
+    std::vector<unsigned int> result;
+    const unsigned int first = this->base_elements.compositional_fields[0];
+    const unsigned int last = *std::max_element(this->base_elements.compositional_fields.begin(), this->base_elements.compositional_fields.end());
+    for (unsigned int i = first; i<=last; ++i)
+      result.emplace_back(i);
+    return result;
+  }
```
Comment (bangerth) [inline source/simulator/introspection.cc:348]: I thought that the function returns the base elements for all compositional fields. Then this array would have size `# of compositional fields`, but it clearly does not here.

File: include/aspect/introspection.h (hunk 1 above)
Comment (tjhei) [reply include/aspect/introspection.h:474]: No, currently it has length 1.

File: source/simulator/introspection.cc (hunk 5 above)
Comment (tjhei) [reply source/simulator/introspection.cc:531]: Oops.

File: source/simulator/introspection.cc (hunk 7 above)
Comment (tjhei) [reply source/simulator/introspection.cc:348]: no, it returns a list of base element indices.

File: source/mesh_refinement/compaction_length.cc (hunk 3 above)
Comment (tjhei) [reply source/mesh_refinement/compaction_length.cc:73]: see #5816

**Comment by tjhei** [comment]
Updated and rebased.

File: source/simulator/introspection.cc (hunk 5 above)
Comment (bangerth) [reply source/simulator/introspection.cc]: OK, so I wasn't totally crazy? :-)

**Review by bangerth — COMMENTED** [review]
Close.

File: source/simulator/introspection.cc [hunk 8]
```diff
@@ -487,6 +524,22 @@ namespace aspect
   }
 
 
+
+  template <int dim>
+  bool
+  Introspection<dim>::is_composition_component (const unsigned int component_index) const
+  {
+    // All compositions live at the end. Just to be sure, there are no other components
+    // in our system after compositional fields, right?
```
Comment (jdannberg) [inline source/simulator/introspection.cc:533]: Does that work in models with melt transport? We do have the melt velocity, fluid pressure etc. before the temperature, right?

File: source/simulator/helper_functions.cc [hunk 9]
```diff
@@ -1746,114 +1766,79 @@ namespace aspect
...
                 {
-                  if (temperature_and_composition_use_same_fe)
-                    distributed_vector(local_dof_indices[temperature_idx]) = in_C.temperature[j];
-                  else
-                    distributed_vector(local_dof_indices[temperature_idx]) = in_T.temperature[j];
+                  // the how-manyth dof is that for the current component and where do we find this in the list
+                  // of evaluation points?
+                  const unsigned int index_within = comp_pair.second;
+                  const unsigned int field_index = component_idx-component_idx_T;
+                  const unsigned int idx = support_point_index_by_field[field_index][index_within];
```
Comment (jdannberg) [inline source/simulator/helper_functions.cc:1826]: I find this hard to follow. I understand the general concept, but if you could add another sentence of explanation (either here, or where you fill the support_point_index_by_field vector) I think that would be helpful.

**Review by jdannberg — APPROVED** [review]
I think this is way better than it was before, I really like it! In a few places, I found it hard to follow what all the different indices are, but it's a clear improvement on my original version.

File: source/simulator/helper_functions.cc [hunk 10]
```diff
@@ -1686,55 +1686,75 @@ namespace aspect
...
-    const Quadrature<dim> quadrature_T(dof_handler.get_fe().base_element(introspection.base_elements.temperature).get_unit_support_points());
+      // Naive n^2 algorithm to merge all compositional fields into the data structure.
+      for (unsigned int c=0; c<introspection.n_compositional_fields; ++c)
+        {
+          const auto composition = Simulator<dim>::AdvectionField::composition(c);
+          const std::vector<Point<dim>> &points = dof_handler.get_fe().base_element(composition.base_element(introspection)).get_unit_support_points();
+
+          for (unsigned int i=0; i<points.size(); ++i)
+            {
+              const auto &it = std::find(unique_support_points.begin(), unique_support_points.end(), points[i]);
```
[the remark this answers is not in the document]
Comment (tjhei) [reply source/simulator/helper_functions.cc:1719]: I think lifetime extension makes this work, but I agree.

File: source/simulator/introspection.cc (hunk 8 above)
Comment (tjhei) [reply source/simulator/introspection.cc:533]: Yes, that is correct. But I added an Assert here just in case.

File: source/simulator/helper_functions.cc (hunk 9 above)
Comment (tjhei) [reply source/simulator/helper_functions.cc:1826]: better?

**Comment by tjhei** [comment]
updated.

**Comment by tjhei** [comment]
Oops, I missed a few comments.

File: source/mesh_refinement/composition_gradient.cc [hunk 11]
```diff
@@ -37,44 +38,51 @@ namespace aspect
...
+                    for (unsigned int j=0; j<dofs_per_cell; ++j)
+                      this_indicator += composition_gradients[j].norm();
+
+                    // Scale gradient in each cell with the correct power of h. Otherwise,
+                    // error indicators do not reduce when refined if there is a density
+                    // jump. We need at least order 1 for the error not to grow when
+                    // refining, so anything >1 should work. (note that the gradient
+                    // itself scales like 1/h, so multiplying it with any factor h^s, s>1
+                    // will yield convergence of the error indicators to zero as h->0)
+                    this_indicator *= std::pow(cell->diameter(), power);
```
[the remark this answers is not in the document]
Comment (tjhei) [reply source/mesh_refinement/composition_gradient.cc:81]: Thank you!

**Comment by tjhei** [comment]
Fingers crossed, this might be good to go now.

**Comment by tjhei** [comment]
The two tests are changing very slightly due to the reordering of the loops in the composition mesh refinement plugin.

File: source/mesh_refinement/composition_gradient.cc [hunk 12]
```diff
@@ -19,6 +19,7 @@
 */
 
 
+#include <aspect/simulator.h>
```
Comment (gassmoeller) [inline source/mesh_refinement/composition_gradient.cc:22]: Ah, we really need to find a solution to put `AdvectionField` out of the `simulator.h` header. Including `simulator.h` in a lot of places is not a clean design. Not for this PR though.

File: source/simulator/helper_functions.cc [hunk 13]
```diff
@@ -1746,114 +1775,88 @@ namespace aspect
...
             {
-              const unsigned int temperature_idx
-                = dof_handler.get_fe().component_to_system_index(introspection.component_indices.temperature,
-                                                                 /*dof index within component=*/ j);
-
-              // skip entries that are not locally owned:
-              if (dof_handler.locally_owned_dofs().is_element(local_dof_indices[temperature_idx]))
+              const auto comp_pair = dof_handler.get_fe().system_to_component_index(dof_idx);
+              const unsigned int component_idx = comp_pair.first;
+              if (component_idx>=component_idx_T) // ignore velocity, pressure, etc.
```
Comment (gassmoeller) [inline source/simulator/helper_functions.cc:1829]: This relies on implicit knowledge about the component ordering. What about:
```suggestion
              if (introspection.is_stokes_component(component_idx) == false) // ignore velocity, pressure, etc.
```

File: source/simulator/helper_functions.cc [hunk 14]
```diff
@@ -1746,114 +1775,88 @@ namespace aspect
...
+                  // field. That means we want to find where this DoF in the computed reactions above to copy it
+                  // back into the global solution vector.
+
+                  // These two variables tell us the how-manyth shape function of which field (and therefore
+                  // field) this DoF is:
+                  const unsigned int index_within = comp_pair.second;
+                  const unsigned int field_index = component_idx-component_idx_T;
+                  // Now we can look up in the support_point_index_by_field data structure where this support
+                  // point is in the list of unique_support_points (and in the Quadrature):
+                  const unsigned int idx = support_point_index_by_field[field_index][index_within];
```
Comment (gassmoeller) [inline source/simulator/helper_functions.cc:1841]:
```suggestion
                  const unsigned int point_idx = support_point_index_by_field[field_index][index_within];
```

File: source/simulator/introspection.cc (hunk 7 above)
Comment (gassmoeller) [reply source/simulator/introspection.cc:349]: I was under the same impression as @bangerth from reading the documentation in the header.

File: source/simulator/introspection.cc [hunk 15]
```diff
@@ -333,6 +333,46 @@ namespace aspect
 
 
 
+  template <int dim>
+  std::vector<unsigned int>
+  Introspection<dim>::get_compositional_field_base_element_indices() const
```
Comment (gassmoeller) [inline source/simulator/introspection.cc:338]: maybe this would help a bit to avoid misunderstanding this function:
```suggestion
  Introspection<dim>::get_composition_base_element_indices() const
```

File: source/simulator/introspection.cc [hunk 16]
```diff
@@ -333,6 +333,46 @@ namespace aspect
...
+  {
+    // We are assigning base elements in order, so the first compositional field
+    // gives us the first base element index. Then we find the largest index
+    // in the vector. This is necessary, because the fields could have type A,B,A.
+    std::vector<unsigned int> result;
+    const unsigned int first = this->base_elements.compositional_fields[0];
+    const unsigned int last = *std::max_element(this->base_elements.compositional_fields.begin(), this->base_elements.compositional_fields.end());
+    for (unsigned int i = first; i<=last; ++i)
+      result.emplace_back(i);
+    return result;
```
Comment (gassmoeller) [inline source/simulator/introspection.cc:348]: Save a few allocations:
```suggestion
    const unsigned int first = this->base_elements.compositional_fields[0];
    const unsigned int last = *std::max_element(this->base_elements.compositional_fields.begin(), this->base_elements.compositional_fields.end());
    std::vector<unsigned int> result(last-first);
    std::iota(result.begin(), result.end(), first);
    return result;
```

Even better: I reworked some parts of introspection to cache this information (see `get_names_for_fields_of_type()`). This way we can avoid allocating a vector in tight loops (the caller of this function might not know it is expensive).

File: source/simulator/introspection.cc [hunk 17]
```diff
@@ -333,6 +333,46 @@ namespace aspect
...
+           ExcMessage("Invalid base_element_index specified."));
+
+    unsigned int idx = 0;
+    for (const auto base_idx : this->base_elements.compositional_fields)
+      {
+        if (base_idx == base_element_index)
+          result.emplace_back(idx);
+
+        ++idx;
+      }
```
Comment (gassmoeller) [inline source/simulator/introspection.cc:368]: Also here, I think it would be better if this could be computed in the constructor once and then just returned by reference (it is called from a loop over all base elements x over all cells). I think of `introspection` as a sort of cache for internal information so it would be ok to precompute things.

**Review by gassmoeller — COMMENTED** [review]
Looks all correct, but could you look into my comments about the documentation? I think the current text is slightly confusing. Also some small suggestions for performance improvements.

File: source/simulator/helper_functions.cc (hunk 13 above)
Comment (tjhei) [reply source/simulator/helper_functions.cc:1829]: Neat idea, but this won't work with melt or other FE modifications, I think.

File: source/mesh_refinement/composition_gradient.cc (hunk 12 above)
Comment (tjhei) [reply source/mesh_refinement/composition_gradient.cc:22]: agreed, see #5887

File: source/simulator/introspection.cc (hunk 17 above)
Comment (tjhei) [reply source/simulator/introspection.cc]: are you ok with a std::map? See the last commit.

**Comment by tjhei** [comment]
updated.

**Comment by tjhei** [comment]
cookbooks/steinberger/

File: source/simulator/introspection.cc (hunk 17 above)
Comment (gassmoeller) [reply source/simulator/introspection.cc]: I am ok with it until it shows up in a profiling run ;-) ... we can keep it for now. But can you change both functions to return references to vectors? Also I think you need to indent.

**Review by gassmoeller — APPROVED** [review]
